### DISCUSSION TITLE
Run all tests of a project

### DISCUSSION THREAD

**Discussion by LorenzoBettini** [mfussenegger/nvim-jdtls#709] [Q&A]
Is there a way to run all tests of a project or of a directory? Something like what you can do in Eclipse.

Neotest and neotest-java provide something like that but it doesn't work very well.

**Comment by mfussenegger** [comment]
See https://github.com/mfussenegger/nvim-jdtls/issues/563#issuecomment-1792245104

**Comment by LorenzoBettini** [comment]
@mfussenegger thank you for the pointer.

I'd like to understand better what's being executed under the hood: from what I understand, the actual launch part is delegated to Eclipse JDT LSP, right? Eclipse itself allows you to run tests from a folder or project. I seem to understand that, instead, the JDT LSP does not provide such a mechanism?

Thanks in advance

**Reply by mfussenegger** [reply]
nvim-jdtls uses the java extension from [vscode-java-test](https://github.com/microsoft/vscode-java-test) for the test discovery, but also for running them.


If you execute a test, it runs a mainclass `org.eclipse.jdt.internal.junit.runner.RemoteTestRunner` with args like `-version 3 -port 39939 -testLoaderClass org.eclipse.jdt.internal.junit4.runner.JUnit4TestLoader -loaderpluginname org.eclipse.jdt.junit4.runtime -test some.example.ClassName:exampleTestMethod`

The `RemoteTestRunner` serves the test output on the `-port`, and nvim-jdtls connects to it to read the results from it.

As mentioned in the other issue, it would possible to discover the test classes, and then run each in a separate debug session.
